cli/goals_principles.md

```diff
@@ -0,0 +1,74 @@
+# Abstract
+
+This document will outline the goals of creating a mobile CLI for K8s / OpenShift. It will also outline the technical considerations and design principles that should be used to guide the development of not only the mobile CLI, but any other mobile service CLI.
+
+
+
+# Goals
+
+The key goal of the mobile cli, is to provide a simple and intuitive interface to enable and assist developers to adopt and use Open Source mobile services on top of Kubernetes and OpenShift in order to solve their mobile project requirements.
+
+
+# General design principles and UX considerations
+
+## Output
+
+As this is a CLI, it is important that it remains a programable grep and pipe friendly interface to allow scripting and execution from external programs. This generally means it should write its output to stdOut. Using a format like JSON allows it to be piped to tools such as [jq](https://stedolan.github.io/jq/) for further use.
+
+However it should default to presenting its output in a human readable format such as a table structure:
+```
++-----------------------+-----------------------------+
+|         NAME          |             ID              |
++-----------------------+-----------------------------+
+| dh-fh-sync-server-apb | dh-fh-sync-server-apb-59dkt |
++-----------------------+-----------------------------+
+```
+
+## Arguments and Flags
+	
+- An argument represents a required parameter needed in order to complete the requested action. 
+- A flag represent an optional value needed for additional functionality or to change default behaviour. Flags can be global, as in spanning all commands or specific to a single command. Were appropriate they should have a short and long format.
+
+```
+mobile get serviceinstances fh-sync-server --output=json
+```
+In this example we have  ```servicesinstances``` and ```fh-sync-server``` that are required but the ```--output``` flag is optional to change the default behaviour for how the output is formatted.
+
+## Errors
+
+In the case of an error, there should be a clear message printed as the first line of output followed by the usage for that command, the process should then exit with a non zero exit code so then when being called from an external program it is clear when something has failed.
```

Comment:
Do you think it might be better to output the human-readable error under the usage instead, to reduce the chance it get's pushed out of the top of the console window?

Reply:
personally no. but I think that perhaps there is an argument for not showing the usage and only showing the error.